feat: Allow external iris for property and resource class when updating doap (DEV-4341)

seakayone opened the pull request:
deprecated: Deprecate endpoints for updating a Default Access Permission in favor of newly introduces `PUT /admin/permissions/doap/:permissionIri` endpoint

Deprecated endpoints:
- `PUT /admin/permissions/:permissionIri/resourceClass`
- `PUT /admin/permissions/:permissionIri/property`
- `PUT /admin/permissions/:permissionIri/group` (deprecated only for DOAPs)
- `PUT /admin/permissions/:permissionIri/hasPermission` (deprecated only for DOAPs)

<!-- Important! Please follow the guidelines for naming Pull Requests: https://docs.dasch.swiss/latest/developers/contribution/ -->

## Pull Request Checklist

### Task Description/Number

<!-- Please add either the issue number or, in case of unscheduled work, a short description of the task at hand -->

Issue Number: DEV-

### PR Type

- [ ] build/chore: maintenance tasks (no production code change)
- [ ] docs: documentation changes (no production code change)
- [x] feat: represents new features
- [ ] fix: represents bug fixes
- [ ] perf: performance improvements
- [ ] refactor: represents production code refactoring
- [ ] test: adding or refactoring tests (no production code change)
- [x] deprecated: Deprecation warning (ideally referencing a migration guide)

### Basic requirements for bug fixes and features

- [ ] Tests for the changes have been added
- [ ] Docs have been added / updated

### Does this PR introduce a breaking change?

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes

jnussbaum commented on webapi/src/main/scala/org/knora/webapi/responders/admin/PermissionsResponder.scala:
```diff
@@ -861,148 +861,54 @@ final case class PermissionsResponder(
+        newForWhat = doap.forWhat match
+                       case Group(_)                       => ResourceClass(resourceClassIri.toInternal)
+                       case ResourceClass(_)               => ResourceClass(resourceClassIri.toInternal)
+                       case Property(p)                    => ResourceClassAndProperty(resourceClassIri.toInternal, p)
+                       case ResourceClassAndProperty(_, p) => ResourceClassAndProperty(resourceClassIri.toInternal, p)
```
I believe I understand your intention, but I guess you are too smart here. I think this behaviour is unintuitive. 

For example there is a DOAP xyz defined for a property, and now I send 

```
PUT 
{
    "forResourceClass": "new-res-class"
}
```

then I end up with this DOAP:

```
{
    "forResourceClass": "new-res-class",
    "forProperty": "old-prop",
}
```

But I argue that's not intended. If I had intended to get that, I would have sent this request in the first place:

```
PUT
{
    "forResourceClass": "new-res-class",
    "forProperty": "old-prop",
}
```

To turn it the other way around: With this PR's implementation, how can I be sure to end up with a DOAP that is solely for a property, but for nothing else? To be sure, I would have to delete the old DOAP and create a new one. 

So I argue to make a "hard overwrite", i.e. delete all the old stuff and add the new stuff.

This behaviour should then also be documented at https://docs.dasch.swiss/latest/DSP-API/03-endpoints/api-admin/permissions/#updating-a-default-object-access-permissions-resource-class

seakayone replied:
The problem is the API is poorly designed for updating the combination of resource class and property:

We have two endpoints which only accept a single property:

`PUT admin/permissions/<doap_permissionIri>/property`
body: `{ "forProperty" : ...}`

`PUT /admin/permissions/<doap_permissionIri>/resourceClass`
body: `{ "forResourceClass" : ...}`

The combination of both parameters in the body - like you suggested - is currently not supported by the existing endpoints. Even the documentation omits the combination of class and property.

jnussbaum replied:
That would be a huge improvement! I don't want to burden too much work on you, but the current situation is really subpar, and could lead to confusion.

seakayone replied:
or create a new endpoint `PUT /admin/permissions/doap/<doap_permissionIri>/forWhat` which would accept every valid combination of group, property and resource class.

So you do not have to choose a different endpoint every time you want to change a doap.

jnussbaum replied:
> PUT /admin/permissions/doap/<doap_permissionIri>/forWhat

Would also be fine for me

seakayone replied:
I believe the old implementation did not create a ResourcClassAndProperty and just replaced it with either the property or resourceclass ForWhat. I will update this PR to be backwards compatible.

seakayone commented:
I have added a new endpoint `PUT /admin/permissions/doap/:permissionIri` which allows for updating an existing doap. This is deprecating existing endpoints as described above. I also updated the documentation.

Please have another review.

jnussbaum reviewed (commented):
Here I already leave some comments. More is to follow.

seakayone commented on docs/03-endpoints/api-admin/permissions.md:
```diff
@@ -268,37 +275,45 @@ Either the `name` or the `permissionCode` must be present; it is not necessary t
+### Updating an existing Default Object Access Permission
+
+- `PUT: /admin/permissions/doap/<doap_permissionIri>` to change the properties of an existing default object 
+access permission, identified by it IRI `<doap_permissionIri>`. 
```
I like them, my formatter seems to break the lines at 80 characters automatically.

seakayone commented on docs/03-endpoints/api-admin/permissions.md:
```diff
@@ -268,37 +275,45 @@ Either the `name` or the `permissionCode` must be present; it is not necessary t
+All properties of the default object access permission are optional and may be combined.
 
-### Updating a Default Object Access Permission's Property
+[NOTE!] Only certain combinations of properties are allowed. The following combinations are allowed:
```
=> bd242aa673436e99273a2be4adc2e43035112c4f

jnussbaum reviewed (commented):
Thanks for your efforts. I have just looked at the documentation. If you think there is a particular code region where I should look at, I would need some directions.

I don't know how our deprecation process looks like. Perhaps you need to inform the frontend team about it?

mpro7 commented on webapi/src/main/scala/org/knora/webapi/slice/admin/api/PermissionsEndpoints.scala:
```diff
@@ -6,16 +6,47 @@
+        forResourceClass.isEmpty &&
+        forProperty.isEmpty &&
+        hasPermissions.isEmpty
+    def hasForWhat: Boolean =
```
For what would this be used? Not quite understand the name here.

mpro7 commented:
Forgot to ask. Should this PR be marked actually as `deprecated` instead of `feat`?

seakayone commented on docs/03-endpoints/api-admin/permissions.md:
```diff
@@ -268,37 +275,45 @@ Either the `name` or the `permissionCode` must be present; it is not necessary t
+
+- `PUT: /admin/permissions/doap/<doap_permissionIri>` to change the attributes of an existing default object 
+access permission, identified by its IRI `<doap_permissionIri>`. 
+The IRI of the new property must be given in the request body as:
```
wdyt 453d55567ee1bf3e00f0da7332b63da988125106 ?

seakayone commented on docs/03-endpoints/api-admin/permissions.md:
```diff
@@ -268,37 +275,45 @@ Either the `name` or the `permissionCode` must be present; it is not necessary t
 {
-    "forResourceClass": "http://api.dasch.swiss/ontology/0803/incunabula/v2#bild"
+  "forProperty" : "http://api.dasch.swiss/ontology/00FF/images/v2#titel",
+  "forResourceClass": "http://api.dasch.swiss/ontology/0803/incunabula/v2#bild",
```
Should be fixed now.

jnussbaum replied:
Hmmm. Still not fixed.

seakayone commented on docs/03-endpoints/api-admin/permissions.md:
```diff
@@ -268,37 +275,45 @@ Either the `name` or the `permissionCode` must be present; it is not necessary t
+If the combination of attributes is not allowed, the request will fail with a `400 Bad Request` error.
+Any valid combination of attributes will replace the existing values. 
+
+If present, the `hasPermissions` property must contain the necessary parameters with respect to the type of the permission 
```
Is this clearer? b7b4c9b0d9043b6ca6319e464929f6be397f32ca

seakayone commented on docs/03-endpoints/api-admin/permissions.md:
```diff
@@ -8,19 +8,20 @@
+| `/admin/permissions/{permissionIri}/hasPermission`          | `PUT`      | [update the scope of an administrative or default object access permission](#updating-a-permissions-scope), i.e. what permissions are granted to which group when this permission applies |
+| `/admin/permissions/{permissionIri}`                        | `DELETE`   | [delete an administrative or default object access permission](#deleting-a-permission)                                                                                                    |
+| ~~`/admin/permissions/{doap_permissionIri}/resourceClass`~~ | `PUT`      | **deprecated**, use `/admin/permissions/doap/{permissionIri}` instead                                                                                                                     |
+| ~~`/admin/permissions/{doap_permissionIri}/property`~~      | `PUT`      | **deprecated**,  use `/admin/permissions/doap/{permissionIri}` instead                                                                                                                    | 
```
Thanks.

jnussbaum commented on docs/03-endpoints/api-admin/permissions.md:
```diff
@@ -222,8 +223,53 @@ Therefore, it is not possible to create new default object access permissions
+      - `forGroup`  <!-- markdownlint-disable MD046 -->
+      - `forResourceClass`
+      - `forProperty`
+      - `forResourceClass` and `forProperty`
```
Before, the unordered list was not rendered correctly by mkdocs. I served locally and played around until it worked, then I committed: https://github.com/dasch-swiss/dsp-api/pull/3425/commits/8f95db8123bac12e999b3b83828267935d2d3d45

seakayone replied:
I see. ` <!-- markdownlint-disable MD046 -->` :/

seakayone replied:
Thanks.

jnussbaum commented on docs/03-endpoints/api-admin/permissions.md:
```diff
@@ -222,8 +223,53 @@ Therefore, it is not possible to create new default object access permissions
+
+All attributes of the default object access permission are optional and may be combined.
+
+!!! note
```
To put more emphasis on it, I would perhaps choose a warning (https://squidfunk.github.io/mkdocs-material/reference/admonitions/#+type:warning)

```suggestion
!!! warning
```

jnussbaum commented on docs/03-endpoints/api-admin/permissions.md:
```diff
@@ -222,8 +223,53 @@ Therefore, it is not possible to create new default object access permissions
+
 ### Updating a Permission's Group
 
+!!! note
```
To put more emphasis on it, I would perhaps choose a warning (https://squidfunk.github.io/mkdocs-material/reference/admonitions/#+type:warning)

```suggestion
!!! warning
```
